### DISCUSSION TITLE
Help with building a new pyodide package with a C library loaded dynamically

### DISCUSSION THREAD

**Discussion by pojungc** [pyodide/pyodide#3570] [Q&A] [answered]
Hi all,

I have been trying to port a python package to WASM. This package depends on a C library, however, the implementation is a bit different than the conventional Python C native package. The C package is packaged into the wheel and it is dynamically loaded during runtime with [ctypes](https://docs.python.org/3/library/ctypes.html#loading-dynamic-link-libraries). The build process is handled by CMake. 

I have been able to build the C library by relying on the [following reference](https://github.com/pyodide/pyodide/issues/3427) ( using the Emscripten's CMake toolchain, with theses flags:`-sSIDE_MODULE`, `-fPIC`,). Which generates a seemingly valid `.so` file. I used [these web assembly tools](https://github.com/WebAssembly/wabt#wabt-the-webassembly-binary-toolkit) to inspect and validate the binary. I don't see any obvious errors and [wasm-validate](https://webassembly.github.io/wabt/doc/wasm-validate.1.html) doesn't return any errors. 

In the build process, the following command is run `pyodide build --export whole_archive` on the Python package and it was able to be successfully built. Further inspection of the wheel, I do see the `.so` file mentioned above.

The issues occurs when loading the package in a web page
`await pyodide.loadPackage("faulty-package-1.0.0-py3-none-any.whl");` and I see the following error related to one of the symbols in the C library:
```
Uncaught (in promise) Error: bad export type for `mypkg_exit_handler`: undefined
    at Object.reportUndefinedSymbols (pyodide.asm.js:10:355375)
    at Object.loadPackage (pyodide.asm.js:10:145161)
    at async main ((index):13:9)
```

Would anyone have an idea of what could be wrong, or how to debug/move forward?

**Comment by hoodmane** [comment]
Could you upload your wheel?

**Reply by pojungc** [reply]
Ah unfortunately, I am not able to :( is there any particular information you were looking for?

**Reply by hoodmane** [reply]
Well I would use `wasm-objdump` to list out the imports and exports for various `.so` files. Presumably something is trying to import `mypkg_exit_handler` but nothing exported it.

**Reply by pojungc** [reply]
Ah I tried `wasm-objdump` but I'm not too familiar in the output, I greped for all the instance of it and found the following,:
```
 - global[5] i32 mutable=1 <- GOT.func. mypkg_exit_handler
 - func[157] sig=2 < mypkg_exit_handler >
 - func[157] < mypkg_exit_handler > -> "mypkg_exit_handler"
 - func[157] size=85 < mypkg_exit_handler >
 - func[157] < mypkg_exit_handler >
 - global[5] < mypkg_exit_handler >
```
Not sure how to read this or what to do with this information. There is only 1 `.so` file

**Reply by hoodmane** [reply]
If there is only one `.so` file then your problem is that your package declares `mypkg_exit_handler` but does not define it. The program assumes that it is going to be dynamically linked in and makes it an input, and it fails at load time when we check that all imports exist.

**Reply by pojungc** [reply]
I am not too sure to understand what you mean, some more additional information: 
- `mypkg_exit_handler` is not declared in the python package, and the only reference to it is in the C library and defined as a function.
- I'm constructing the `.so` file using CMake's `add_library()` call and it generated with `emcc` with a bunch of `.a` files that the built C library depends on, are you saying that we need to include all the `.a` files in our wheel as well?

**Reply by hoodmane** [reply]
No the `.a` files are archives, they are presumably linked into the final `.so` file and certainly should not be included in the final output. Which source file is `mypkg_exit_handler` defined in? Presumably the object file for that compilation unit wasn't linked in for some reason? Can you see other symbols defined in the same source file or are they missing too?

**Comment by ryanking13** [comment]
Does the C-library included in the wheel? How do you locate and load that library in the Python code?

**Reply by pojungc** [reply]
In the build process there is logic to generate the python bindings of the C libraries and that is included in the wheel as well as the generated dynamic C library. The dll is located in the wheel and loaded with the ctypes library:  https://docs.python.org/3/library/ctypes.html#loading-dynamic-link-libraries

**Comment by pojungc** [comment, marked as answer]
Found some logic related to file extension, and as an experiment I renamed the `.so` extension to a random extension and packaged it into the wheel. I was able to get past this issue! 
Although I am hitting another issue. 
```
RangeError: WebAssembly.Compile is disallowed on the main thread, if the buffer size is larger than 4KB. Use WebAssembly.compile, or compile on a worker thread.
```
As an experiment I changed the C library to a custom one less than 4kb and that seems to work as expected, so at least that indicates that the wheel is somewhat okay. I will continue digging into a solution for above and mark this as resolved for now.
Thanks @hoodmane and @ryanking13 for looking into this question!

**Reply by hoodmane** [reply]
The problem is that the extension needs to be `.so` for Pyodide to load it. The logic is here:
https://github.com/pyodide/pyodide/blob/main/src/py/pyodide/_package_loader.py#L226
The file name must match `\.so(.\d+)*$` (`SHAREDLIB_REGEX`). 

Maybe once the wheel is generated you can unpack it with `python -m wheel unpack`, change the shared library extension back to `.so`, and then repack it with `python -m wheel pack`.
